### PR TITLE
Fix UDS Tests, Improve MCU Makefile rules, Handle OTA Edge Case & Enhance Logs

### PR DESCRIPTION
**Pull request by AlexZahariaRandstad**
## Description

This MR fixes:

- All tests from the UDS module, with some requiring a rewrite for correctness, with all needed modifications for setters, getters, and additional utility functions.
- Improvements to the MCU Makefile, adding runUdsTests and cleanUdsTests rules.
- Add a fix for handling cases where OTA is already initialized.
- Additionally, I added documentation and meaningful logs to improve clarity and debugging.

Steps to Test:

1. In MCU terminal, run: `make runUdsTests` to compile all tests for UDS and run them.
image

2. As a result, all tests should pass.
image

3. To clean all generated files, run `make cleanUdsTests`.
image

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: mcu/Makefile [hunk 1]
```diff
@@ -482,6 +485,16 @@ $(OBJ_DIR)/NegativeResponse_test.o: $(UTILS_DIR)/NegativeResponse.cpp
...
+# Compile and Run all UDS unit tests
+runUdsTests: runSecurityAccessTest runAccessTimingParameterTests runClearDtcTests \
+	runDiagnosticSessionControlTests runEcuResetTests runReadDataByIdentifierTests runReadDtcTests \
+	runRoutineControlTests runTesterPresentTests runWriteDataByIdentifierTests
+
+# Clean all UDS unit tests
+cleanUdsTests: cleanSecurityAccessTest cleanAccessTimingParameterTests cleanClearDtcTests \
+	cleanDiagnosticSessionControlTests cleanEcuResetTests cleanReadDataByIdentifierTests \
+	cleanReadDtcTests cleanRoutineControlTests cleanTesterPresentTests cleanWriteDataByIdentifierTests \
+	clean
```
[the remark this answers is not in the document]
Comment (AlexZahariaRandstad) [reply mcu/Makefile:497]: The clean target in the MCU makefile is configured to clean only the files generated during the MCU build process, such as object files, MCU-specific executables, and other intermediate files created during the compilation and linking process at the MCU level. It does not affect the files generated for unit tests, which are part of a different build process.
Files associated with unit tests (such as .out, .o, .gcno, .gcda) are not included in the MCU build process and, therefore, the clean target in the MCU makefile will not automatically delete them.
To clean these files and maintain a clean development environment between test runs, it is necessary to add a separate clean target specifically for unit tests. This target will remove "noise" files, ensuring that no artifacts interfere with the new test executions.
Therefore, to clean the files generated by unit tests, a separate target, such as cleanUdsTests, needs to be defined. This target will delete test-specific files from their respective directories.

File: ota/request_download/src/RequestDownload.cpp [hunk 2]
```diff
@@ -418,4 +418,9 @@ bool RequestDownloadService::isLatestSoftwareVersion()
...
 {
     return RequestDownloadService::rds_data;
-}
\ No newline at end of file
+}
+
+void RequestDownloadService::vSetRdsDataFormat(uint8_t u8Format)
+{
+    RequestDownloadService::rds_data.data_format = u8Format;
+}
```
[the remark this answers is not in the document]
Comment (AlexZahariaRandstad) [reply ota/request_download/src/RequestDownload.cpp:426]: The idea is good, but it's not the best fit in this case. The current code isn't very user-friendly for dependency injection, which means controlling it with a mock would be difficult. Additionally, the main modifications are tied to processFrameData, and making changes there risks breaking more than fixing. Achieving that would require a lot of code refactoring and testing, which isn't considered worth investing in at this point in the project.

File: mcu/Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply mcu/Makefile:497]: "The clean target in the MCU makefile is configured to clean only the files generated during the MCU build process..." not quite. As far as I have seen/used the make clean, I can delete any test/coverage artifact. Please have a look at this image
Screenshot from 2025-02-21 10-47-34

File: ota/request_download/src/RequestDownload.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply ota/request_download/src/RequestDownload.cpp:426]: "The current code isn't very user-friendly for dependency injection" I agree with you on this. Please add comments stating that this method should be test only and deleted as soon as we refactor the code to be better suited for dependency injection (followup tickets to be opened)

File: uds/read_dtc_information/include/ReadDtcInformation.h [hunk 3]
```diff
@@ -47,6 +47,13 @@ class ReadDTC
          */
         void read_dtc(int id, std::vector<uint8_t> data);
 
+        /**
+         * @brief Sets the path to the specified folder.
+         * 
+         * @param strPathFolder The path to the target folder.
+         */
+        void vSetPath(std::string strPathFolder);
+
```
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply uds/read_dtc_information/include/ReadDtcInformation.h:56]: Done, but the method is still here?

File: uds/write_data_by_identifier/utest/WriteDataByIdentifierTest.cpp [hunk 4]
```diff
@@ -40,6 +41,18 @@ struct WriteDataByIdentifierTest : testing::Test
...
+        auto* pVecU16ModifiableDIDs = const_cast<std::vector<uint16_t>*>(&vecU16WritableDIDs);
+
+        auto itVecU16Did = std::find(pVecU16ModifiableDIDs->begin(), pVecU16ModifiableDIDs->end(), u16OldDid);
+        if (itVecU16Did != pVecU16ModifiableDIDs->end()) {
+            *itVecU16Did = u16NewDid;
+        } else {
+            pVecU16ModifiableDIDs->push_back(u16NewDid);
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply uds/write_data_by_identifier/utest/WriteDataByIdentifierTest.cpp]: Extracted as FileManager method. It is a valid choice, however, you did not answer the thread

File: uds/read_dtc_information/utest/ReadDtcInformationTest.cpp [hunk 5]
```diff
@@ -50,41 +55,46 @@ struct ReadDtcTest : testing::Test
...
+        }
+
+        ofOutFile << "P0190 24\n";
+        ofOutFile << "P0196 24\n";
+        ofOutFile << "P0069 24\n";
+        ofOutFile << "P0115 24\n";
+
+        ofOutFile.close();
+    }
+
```
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply uds/read_dtc_information/utest/ReadDtcInformationTest.cpp]: You extracted this test related function as a method inside of a production class (FileManager). I do not support this change

File: mcu/Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (AlexZahariaRandstad) [reply mcu/Makefile:497]: Yes, you are right. But the make clean target in the MCU Makefile should have only one responsibility: to remove only the files generated during the MCU build process and should not affect the entire test suite.

File: uds/read_dtc_information/include/ReadDtcInformation.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexZahariaRandstad) [reply uds/read_dtc_information/include/ReadDtcInformation.h]: I hadn't realized it was still there. Done

File: mcu/Makefile (hunk 1 above)
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply mcu/Makefile:497]: So, as far as I understand, you seek to add new cleaning targets. While not a high prioritary task, I do support the idea. However, this could be a separate ticket, creating cleaning tasks for everything, and renaming the current clean to cleanAll.

File: uds/read_dtc_information/utest/ReadDtcInformationTest.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexZahariaRandstad) [reply uds/read_dtc_information/utest/ReadDtcInformationTest.cpp]: Done. Moved in TestUtils.

File: ota/request_download/src/RequestDownload.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (StefanCorneteanu) [reply ota/request_download/src/RequestDownload.cpp:426]: Please do not ignore the last comment! ("add comments stating that this method should be test only and deleted as soon as we refactor the code to be better suited for dependency injection (followup tickets to be opened)"). Would go well to the Doxygen in the header file

File: utils/include/FileManager.h [hunk 6]
```diff
@@ -125,6 +134,17 @@ class FileManager
...
+     * 
+     * @param vecU16WritableDIDs A vector containing writable DID values
+     * @param u16OldDid The DID to be replaced
+     * @param u16NewDid The new DID to insert in place of the old one
+     */
+    static void vReplaceOrAddWritableDID(const std::vector<uint16_t>& vecU16WritableDIDs, uint16_t u16OldDid, uint16_t u16NewDid);
+
 };
 
 #endif
```
Comment (StefanCorneteanu) [inline utils/include/FileManager.h:150]: Add a new line to avoid Github warning

File: uds/routine_control/utest/RoutineControlTest.cpp [hunk 7]
```diff
@@ -71,11 +74,34 @@ struct RoutineControlTest : testing::Test
...
+            throw std::runtime_error("Failed to create fake ZIP file: " + strFilePath);
+        }
+
+        ofOutFile.write(reinterpret_cast<const char*>(vecU8ZipSignature.data()), vecU8ZipSignature.size());
+
+        std::vector<uint8_t> vecU8DummyContent = {'T', 'E', 'S', 'T'};
+        ofOutFile.write(reinterpret_cast<const char*>(vecU8DummyContent.data()), vecU8DummyContent.size());
+
+        ofOutFile.close();
+    }
```
[the remark this answers is not in the document]
Comment (AlexZahariaRandstad) [reply uds/routine_control/utest/RoutineControlTest.cpp]: Now, it is really done.

File: ota/request_download/src/RequestDownload.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexZahariaRandstad) [reply ota/request_download/src/RequestDownload.cpp:426]: Done

File: mcu/Makefile [hunk 8]
```diff
@@ -69,6 +69,8 @@ HVAC_DIR_TEST = ../ecu_simulation/HVACModule/test
 all:
 	$(MAKE) -C ../config init
 
+configInit:
+	$(MAKE) -C ../config init
```
Comment (StefanCorneteanu) [inline mcu/Makefile:73]: I am currently treating this as part of https://trello.com/c/AxYZ0ERf/84-configini-created-only-by-make-all. Please remove any reference to configInit target

File: utils/include/TestUtils.h [hunk 9]
```diff
@@ -95,4 +99,34 @@ int createSocket(uint8_t interface_number);
...
+    explicit FileGuard(const std::string& filePath) : m_filePath(filePath) {}
+    ~FileGuard() {
+        if (std::filesystem::exists(m_filePath)) {
+            std::filesystem::remove(m_filePath);
+        }
+    }
+
+private:
+    std::string m_filePath;
+};
```
Comment (StefanCorneteanu) [inline utils/include/TestUtils.h:130]: I do like the concept of this class, and we might expand on it based on the needs of the tests. Might be helpful to extract as a standalone class

File: uds/clear_dtc/utest/ClearDtcTest.cpp [hunk 10]
```diff
@@ -88,8 +90,14 @@ TEST_F(ClearDtcTest, InvalidGroup) {
 /* Test for Clear Body DTCs */
 TEST_F(ClearDtcTest, ClearBodyDTCs) {
     std::cerr << "Running ClearBodyDTCs" << std::endl;
-
-    struct can_frame result_frame = createFrame(0x10FA, {0x01, 0x54});
+    const std::map<std::string, std::vector<uint16_t>> mapStrU16DtcValues = {
```
Comment (VladVestemeanuRandstad) [inline uds/clear_dtc/utest/ClearDtcTest.cpp:93]: mapStrVecU16_DtcValues, same in other places

File: uds/clear_dtc/utest/ClearDtcTest.cpp [hunk 11]
```diff
@@ -88,8 +90,14 @@ TEST_F(ClearDtcTest, InvalidGroup) {
...
+        {"P0190", {0x24}},
+        {"P0115", {0x24}}
+    };
+    // Instantiate the guard to ensure file deletion on scope exit
+    FileGuard fileGuard(strDtcFilePath);
+    vCreateDummyDtcFile(strDtcFilePath, mapStrU16DtcValues);
+    struct can_frame result_frame = createFrame(0x10FA, {0x02, 0x54, 0x00});
```
Comment (VladVestemeanuRandstad) [inline uds/clear_dtc/utest/ClearDtcTest.cpp:100]: resultFrame

File: uds/read_data_by_identifier/utest/ReadDataByIdentifierTest.cpp [hunk 12]
```diff
@@ -220,10 +224,30 @@ TEST_F(ReadDataByIdentifierTest, ErrorReadingFromFile)
 /* Test Corect DID MCU */
 TEST_F(ReadDataByIdentifierTest, CorectDIDMCU)
 {
-    struct can_frame result_frame = createFrame(0x10FA, {0x04, 0x62, 0xF1, 0xA2, 0x00});
+    std::unordered_map<uint16_t, std::vector<uint8_t>> mapU16VecU8Data = {
```
Comment (VladVestemeanuRandstad) [inline uds/read_data_by_identifier/utest/ReadDataByIdentifierTest.cpp:227]: Similar with an underscore, same in other places

File: utils/include/FileManager.h [hunk 13]
```diff
@@ -125,6 +134,17 @@ class FileManager
...
+
+
+    /**
+     * @brief Replace or add a writable DID
+     * 
+     * @param vecU16WritableDIDs A vector containing writable DID values
+     * @param u16OldDid The DID to be replaced
+     * @param u16NewDid The new DID to insert in place of the old one
+     */
+    static void vReplaceOrAddWritableDID(const std::vector<uint16_t>& vecU16WritableDIDs, uint16_t u16OldDid, uint16_t u16NewDid);
```
Comment (VladVestemeanuRandstad) [inline utils/include/FileManager.h:146]: v_ReplaceOrAddWritableDID()

File: utils/include/FileManager.h [hunk 14]
```diff
@@ -43,6 +44,14 @@ class FileManager
      */
     static std::unordered_map<uint16_t, std::string> getExistingDIDValues(const std::string& strFilePath);
 
+    /**
+     * @brief Method used to append unique key-value pairs from a map to a file.
+     * 
+     * @param strFileName The name of the file where the data will be appended (if not exist).
+     * @param mapU16VecU8Data The map containing key-value pairs to be written to the file.
+     */
+    static void vAppendUniqueMapToFile(const std::string& strFileName, const std::unordered_map<uint16_t, std::vector<uint8_t>>& mapU16VecU8Data);
```
Comment (VladVestemeanuRandstad) [inline utils/include/FileManager.h:53]: v_AppendUniqueMapToFile()

File: utils/include/TestUtils.h [hunk 15]
```diff
@@ -95,4 +99,34 @@ int createSocket(uint8_t interface_number);
...
+ * @param mapStrU16DtcValues A map containing DTC codes
+ */
+void vCreateDummyDtcFile(const std::string& strFilePath, const std::map<std::string, std::vector<uint16_t>>& mapStrU16DtcValues);
+
+/**
+ * @brief Create a dummy ZIP file
+ *
+ * @param strFilePath The path where the dummy ZIP file will be created
+ */
+void vCreateDummyZipFile(const std::string& strFilePath);
```
Comment (VladVestemeanuRandstad) [inline utils/include/TestUtils.h:116]: Similar with underscore for these

File: utils/src/FileManager.cpp [hunk 16]
```diff
@@ -5,6 +5,71 @@
...
+
+    auto itVecU16Did = std::find(pVecU16ModifiableDIDs->begin(), pVecU16ModifiableDIDs->end(), u16OldDid);
+    if (itVecU16Did != pVecU16ModifiableDIDs->end()) {
+        *itVecU16Did = u16NewDid;
+    } else {
+        pVecU16ModifiableDIDs->push_back(u16NewDid);
+    }
+}
+
+void FileManager::vAppendUniqueMapToFile(const std::string& strFileName, const std::unordered_map<uint16_t, std::vector<uint8_t>>& mapU16VecU8Data)
```
Comment (VladVestemeanuRandstad) [inline utils/src/FileManager.cpp:23]: Similar with underscore for these

File: uds/clear_dtc/include/ClearDtc.h [hunk 17]
```diff
@@ -43,6 +43,13 @@ class ClearDtc
          */
         void clearDtc(int id, std::vector<uint8_t> data);
 
+        /**
+         * @brief Retrieves the path to the DTC storage file.
+         *
+         * @return std::string The file path where DTC data is stored.
+         */
+        std::string strGetPathToDTC() const;
```
Comment (VladVestemeanuRandstad) [inline uds/clear_dtc/include/ClearDtc.h:51]: str_GetPathToDTC()

File: uds/clear_dtc/utest/ClearDtcTest.cpp (hunk 10 above)
Comment (AlexZahariaRandstad) [reply uds/clear_dtc/utest/ClearDtcTest.cpp]: Fixed

File: uds/clear_dtc/utest/ClearDtcTest.cpp (hunk 11 above)
Comment (AlexZahariaRandstad) [reply uds/clear_dtc/utest/ClearDtcTest.cpp]: Fixed

File: utils/src/FileManager.cpp (hunk 16 above)
Comment (AlexZahariaRandstad) [reply utils/src/FileManager.cpp]: Done

File: utils/src/FileManager.cpp [hunk 18]
```diff
@@ -5,6 +5,71 @@
 
 #include "FileManager.h"
 #include "Globals.h"
+#include <set>
+#include <map>
+
+
+void FileManager::v_ReplaceOrAddWritableDID(const std::vector<uint16_t>& vecU16WritableDIDs, uint16_t u16OldDid, uint16_t u16NewDid) {
+    auto* pVecU16ModifiableDIDs = const_cast<std::vector<uint16_t>*>(&vecU16WritableDIDs);
```
Comment (VladVestemeanuRandstad) [inline utils/src/FileManager.cpp:13]: There should be an underline between type and name when the attribute/container/structure is more complex. "pVecU16_ModifiableDIDs"

File: utils/src/FileManager.cpp [hunk 19]
```diff
@@ -5,6 +5,71 @@
...
+
+    auto itVecU16Did = std::find(pVecU16ModifiableDIDs->begin(), pVecU16ModifiableDIDs->end(), u16OldDid);
```
Comment (VladVestemeanuRandstad) [inline utils/src/FileManager.cpp:15]: Same

File: utils/src/FileManager.cpp [hunk 20]
```diff
@@ -5,6 +5,71 @@
...
+    if (itVecU16Did != pVecU16ModifiableDIDs->end()) {
+        *itVecU16Did = u16NewDid;
+    } else {
+        pVecU16ModifiableDIDs->push_back(u16NewDid);
+    }
+}
+
+void FileManager::v_AppendUniqueMapToFile(const std::string& strFileName, const std::unordered_map<uint16_t, std::vector<uint8_t>>& mapU16VecU8Data)
+{
+    std::unordered_map<uint16_t, std::vector<uint8_t>> mapU16VecU8ExistingData;
```
Comment (VladVestemeanuRandstad) [inline utils/src/FileManager.cpp:25]: Same

File: utils/src/FileManager.cpp [hunk 21]
```diff
@@ -5,6 +5,71 @@
...
+
+    std::ifstream ifInFile(strFileName);
+    if (ifInFile.is_open())
+    {
+        std::string strLine;
+        while (std::getline(ifInFile, strLine))
+        {
+            std::istringstream iss(strLine);
+            uint16_t u16ExistingDid;
+            std::vector<uint8_t> vecU8ExistingValue;
```
Comment (VladVestemeanuRandstad) [inline utils/src/FileManager.cpp:35]: Same

File: utils/src/TestUtils.cpp [hunk 22]
```diff
@@ -12,6 +12,59 @@
...
+#include <map>
+
+void v_CreateDummyZipFile(const std::string& strFilePath)
+{
+    if (std::filesystem::exists(strFilePath))
+    {
+        std::filesystem::remove(strFilePath);
+    }
+
+    std::vector<uint8_t> vecU8ZipSignature = {0x50, 0x4B, 0x03, 0x04};
```
Comment (VladVestemeanuRandstad) [inline utils/src/TestUtils.cpp:26]: Same

File: utils/src/TestUtils.cpp [hunk 23]
```diff
@@ -12,6 +12,59 @@
...
+
+    std::ofstream ofOutFile(strFilePath, std::ios::binary | std::ios::trunc);
+    if (!ofOutFile.is_open())
+    {
+        throw std::runtime_error("Failed to create dummy ZIP file: " + strFilePath);
+    }
+
+    ofOutFile.write(reinterpret_cast<const char*>(vecU8ZipSignature.data()), vecU8ZipSignature.size());
+
+    std::vector<uint8_t> vecU8DummyContent = {'T', 'E', 'S', 'T'};
```
Comment (VladVestemeanuRandstad) [inline utils/src/TestUtils.cpp:36]: Same

File: uds/read_dtc_information/utest/ReadDtcInformationTest.cpp [hunk 24]
```diff
@@ -125,6 +124,50 @@ TEST_F(ReadDtcTest, NoDtcFound)
...
+        {"P0190", {0x24}},
+        {"P0196", {0x24}},
+        {"P0069", {0x24}},
+        {"P0115", {0x24}}
+    };
+    // Instantiate the guard to ensure file deletion on scope exit
+    FileGuard fileGuard(strPathDummy);
+    v_CreateDummyDtcFile(strPathDummy, mapStrVecU16_DtcValues);
+    r = new ReadDTC(*logger, strPathDummy,socket2_);
+    struct can_frame result_frame = createFrame(0x10fa, {0x010, 0x0f, 0x59, 0x02, 0x24, 0x00, 0x69, 0x24});
```
Comment (VladVestemeanuRandstad) [inline uds/read_dtc_information/utest/ReadDtcInformationTest.cpp:160]: resultFrame

File: uds/read_dtc_information/utest/ReadDtcInformationTest.cpp [hunk 25]
```diff
@@ -125,6 +124,50 @@ TEST_F(ReadDtcTest, NoDtcFound)
...
+        {"P0190", {0x24}},
+        {"P0196", {0x24}},
+        {"P0069", {0x24}},
+        {"P0115", {0x24}}
+    };
+    // Instantiate the guard to ensure file deletion on scope exit
+    FileGuard fileGuard(strPathDummy);
+    v_CreateDummyDtcFile(strPathDummy, mapStrVecU16_DtcValues);
+    r = new ReadDTC(*logger, strPathDummy,socket2_);
+    struct can_frame result_frame = createFrame(0x12fa, {0x06, 0x59, 0x01, 0x24, 0x01, 0x00, 0x04});
```
Comment (VladVestemeanuRandstad) [inline uds/read_dtc_information/utest/ReadDtcInformationTest.cpp:140]: resultFrame

**Review by VladVestemeanuRandstad — APPROVED** [review]

File: ota/transfer_data/src/TransferData.cpp [hunk 26]
```diff
@@ -97,7 +97,7 @@ void TransferData::processDataForTransfer(canid_t can_id, std::vector<uint8_t>&
         LOG_INFO(logger.GET_LOGGER(), "Data processing for transfer started.");
         return;
     }
-
```
Comment (VladVestemeanuRandstad) [inline ota/transfer_data/src/TransferData.cpp:100]: This can be reverted. It will be handled with Clang Format

**Review by StefanCorneteanu — APPROVED** [review]
